presto-memory/src/test/java/io/prestosql/plugin/memory/TestMemorySmoke.java

```diff
@@ -263,6 +263,81 @@ public void testSemiJoinDynamicFilteringBlockProbeSide()
                 ImmutableSet.of(1, ORDERS_COUNT, PART_COUNT));
     }
 
+    @Test
+    public void testCrossJoinDynamicFiltering()
+    {
+        // "<" pushdown
+        {
+            assertDynamicFiltering(
+                    "SELECT * FROM lineitem JOIN orders ON " +
+                            "lineitem.orderkey < orders.orderkey + 1 AND " +
```

Comment:
IIUC, it worked since the optimizer replaced `orderkey + 1` with an `expr` symbol by adding a projection.
I will rewrite the tests to use single symbols in the comparison expressions.

Reply:
> IIUC, it worked since the optimizer replaced orderkey + 1 

Why it replaced it with `expr`?

> I will rewrite the tests to use single symbols in the comparison expressions.

We should test both type of comparisons to document current behavior.

Reply:
It seems that current `master` behaves this way (i.e. adds a projection for cross-join build-side expression) - reproduced with https://github.com/prestosql/presto/compare/master...rzeyde-varada:repro-cross-join-projection:
```java
    @Test
    public void testAddProjectForComplexBuildExpression()
    {
        assertPlan("SELECT 1 FROM lineitem JOIN orders ON lineitem.orderkey < orders.orderkey + 1",
                anyTree(
                        filter("L_ORDERKEY < EXPR",
                                join(INNER, ImmutableList.of(), Optional.empty(),
                                        tableScan("lineitem", ImmutableMap.of("L_ORDERKEY", "orderkey")),
                                        exchange(project(
                                                ImmutableMap.of("EXPR", expression("O_ORDERKEY + BIGINT '1'")),
                                                tableScan("orders", ImmutableMap.of("O_ORDERKEY", "orderkey"))))))));
    }
```

Reply:
IIUC, the initial plan (before the optimization) seems to contain the projection - so it seems that running the optimizers doesn't remove it:
```
Output[_col0]
│   Layout: [expr_3:integer]
│   _col0 := expr_3
└─ Project[]
   │   Layout: [orderkey:bigint, partkey:bigint, suppkey:bigint, linenumber:integer, quantity:double, extendedprice:double, discount:double, tax:double, returnflag:varchar(1), linestatus:varchar(1), shipdate:date, commitdate:date, receiptdate:date, shipinstruct:varchar(25), shipmode:varchar(10), comment:varchar(44), row_number:bigint, orderkey_0:bigint, custkey:bigint, orderstatus:varchar(1), totalprice:double, orderdate:date, orderpriority:varchar(15), clerk:varchar(15), shippriority:integer, comment_1:varchar(79), row_number_2:bigint, expr:bigint, expr_3:integer]
   └─ Project[]
      │   Layout: [orderkey:bigint, partkey:bigint, suppkey:bigint, linenumber:integer, quantity:double, extendedprice:double, discount:double, tax:double, returnflag:varchar(1), linestatus:varchar(1), shipdate:date, commitdate:date, receiptdate:date, shipinstruct:varchar(25), shipmode:varchar(10), comment:varchar(44), row_number:bigint, orderkey_0:bigint, custkey:bigint, orderstatus:varchar(1), totalprice:double, orderdate:date, orderpriority:varchar(15), clerk:varchar(15), shippriority:integer, comment_1:varchar(79), row_number_2:bigint, expr:bigint, expr_3:integer]
      └─ Project[]
         │   Layout: [orderkey:bigint, partkey:bigint, suppkey:bigint, linenumber:integer, quantity:double, extendedprice:double, discount:double, tax:double, returnflag:varchar(1), linestatus:varchar(1), shipdate:date, commitdate:date, receiptdate:date, shipinstruct:varchar(25), shipmode:varchar(10), comment:varchar(44), row_number:bigint, orderkey_0:bigint, custkey:bigint, orderstatus:varchar(1), totalprice:double, orderdate:date, orderpriority:varchar(15), clerk:varchar(15), shippriority:integer, comment_1:varchar(79), row_number_2:bigint, expr:bigint, expr_3:integer]
         └─ FilterProject[filterPredicate = ("orderkey" < "expr")]
            │   Layout: [orderkey:bigint, partkey:bigint, suppkey:bigint, linenumber:integer, quantity:double, extendedprice:double, discount:double, tax:double, returnflag:varchar(1), linestatus:varchar(1), shipdate:date, commitdate:date, receiptdate:date, shipinstruct:varchar(25), shipmode:varchar(10), comment:varchar(44), row_number:bigint, orderkey_0:bigint, custkey:bigint, orderstatus:varchar(1), totalprice:double, orderdate:date, orderpriority:varchar(15), clerk:varchar(15), shippriority:integer, comment_1:varchar(79), row_number_2:bigint, expr:bigint, expr_3:integer]
            │   expr_3 := 1
            └─ CrossJoin
               │   Layout: [orderkey:bigint, partkey:bigint, suppkey:bigint, linenumber:integer, quantity:double, extendedprice:double, discount:double, tax:double, returnflag:varchar(1), linestatus:varchar(1), shipdate:date, commitdate:date, receiptdate:date, shipinstruct:varchar(25), shipmode:varchar(10), comment:varchar(44), row_number:bigint, orderkey_0:bigint, custkey:bigint, orderstatus:varchar(1), totalprice:double, orderdate:date, orderpriority:varchar(15), clerk:varchar(15), shippriority:integer, comment_1:varchar(79), row_number_2:bigint, expr:bigint]
               ├─ Project[]
               │  │   Layout: [orderkey:bigint, partkey:bigint, suppkey:bigint, linenumber:integer, quantity:double, extendedprice:double, discount:double, tax:double, returnflag:varchar(1), linestatus:varchar(1), shipdate:date, commitdate:date, receiptdate:date, shipinstruct:varchar(25), shipmode:varchar(10), comment:varchar(44), row_number:bigint]
               │  └─ ScanProject[table = local:lineitem:sf0.01]
               │         Layout: [orderkey:bigint, partkey:bigint, suppkey:bigint, linenumber:integer, quantity:double, extendedprice:double, discount:double, tax:double, returnflag:varchar(1), linestatus:varchar(1), shipdate:date, commitdate:date, receiptdate:date, shipinstruct:varchar(25), shipmode:varchar(10), comment:varchar(44), row_number:bigint]
               │         orderkey := tpch:orderkey
               │         partkey := tpch:partkey
               │         suppkey := tpch:suppkey
               │         linenumber := tpch:linenumber
               │         quantity := tpch:quantity
               │         extendedprice := tpch:extendedprice
               │         discount := tpch:discount
               │         tax := tpch:tax
               │         returnflag := tpch:returnflag
               │         linestatus := tpch:linestatus
               │         shipdate := tpch:shipdate
               │         commitdate := tpch:commitdate
               │         receiptdate := tpch:receiptdate
               │         shipinstruct := tpch:shipinstruct
               │         shipmode := tpch:shipmode
               │         comment := tpch:comment
               │         row_number := tpch:row_number
               └─ Project[]
                  │   Layout: [orderkey_0:bigint, custkey:bigint, orderstatus:varchar(1), totalprice:double, orderdate:date, orderpriority:varchar(15), clerk:varchar(15), shippriority:integer, comment_1:varchar(79), row_number_2:bigint, expr:bigint]
                  └─ ScanProject[table = local:orders:sf0.01]
                         Layout: [orderkey_0:bigint, custkey:bigint, orderstatus:varchar(1), totalprice:double, orderdate:date, orderpriority:varchar(15), clerk:varchar(15), shippriority:integer, comment_1:varchar(79), row_number_2:bigint, expr:bigint]
                         expr := ("orderkey_0" + CAST(1 AS bigint))
                         orderkey_0 := tpch:orderkey
                         custkey := tpch:custkey
                         orderstatus := tpch:orderstatus
                             :: [[F], [O], [P]]
                         totalprice := tpch:totalprice
                         orderdate := tpch:orderdate
                         orderpriority := tpch:orderpriority
                         clerk := tpch:clerk
                         shippriority := tpch:shippriority
                         comment_1 := tpch:comment
                         row_number_2 := tpch:row_number
```
(printed via https://github.com/prestosql/presto/commit/bc72a40a4512afb7949148df01db80a18e47dca0)

Reply:
BTW, it seems that the same happens when the complex expression is on the probe-side of the join, i.e.:
```sql
SELECT lineitem.orderkey, orders.orderkey 
FROM lineitem JOIN orders 
ON lineitem.orderkey + 1 < orders.orderkey + 1
```
The resulting plan is:
```
Output[orderkey, orderkey]
│   Layout: [orderkey:bigint, orderkey_0:bigint]
│   orderkey := orderkey_0
└─ FilterProject[filterPredicate = ("expr" < "expr_3")]
   │   Layout: [orderkey:bigint, orderkey_0:bigint]
   └─ CrossJoin
      │   Layout: [orderkey:bigint, expr:bigint, orderkey_0:bigint, expr_3:bigint]
      ├─ ScanProject[table = local:lineitem:sf0.01]
      │      Layout: [orderkey:bigint, expr:bigint]
      │      expr := ("orderkey" + BIGINT '1')
      │      orderkey := tpch:orderkey
      └─ LocalExchange[SINGLE] ()
         │   Layout: [orderkey_0:bigint, expr_3:bigint]
         └─ ScanProject[table = local:orders:sf0.01]
                Layout: [orderkey_0:bigint, expr_3:bigint]
                expr_3 := ("orderkey_0" + BIGINT '1')
                orderkey_0 := tpch:orderkey
                tpch:orderstatus
                    :: [[F], [O], [P]]
```

Reply:
I've created https://github.com/prestosql/presto/issues/5777